File: aat-lib/src/main/java/ch/bailu/aat_lib/search/poi/PoiListItem.java
package ch.bailu.aat_lib.search.poi;

import org.mapsforge.poi.storage.PoiCategory;

import ch.bailu.aat_lib.lib.filter_list.KeyList;
import ch.bailu.aat_lib.lib.filter_list.ListEntry;

public class PoiListItem extends ListEntry {
    private final PoiCategory self;

    private final boolean isSummary;
    private boolean isSelected = false;

    private final KeyList keys;
    private final String summaryKey;


    public PoiListItem(PoiCategory c) {
        isSummary = true;
        self = c;

        keys = new KeyList(c.getTitle());
        summaryKey = "_" + keys.getKey(0);
        keys.addKeys(summaryKey);
    }

    public PoiListItem(PoiCategory c, PoiListItem s) {
        isSummary = false;

        self = c;

        summaryKey = s.getSummaryKey();
        keys = new KeyList(c.getTitle());
        keys.addKeys(summaryKey);
    }

    @Override
    public boolean isSelected() {
        return isSelected;
    }

    @Override
    public boolean isSummary() {
        return isSummary;
    }

    @Override
    public KeyList getKeys() {
        return keys;
    }

    @Override
    public String getSummaryKey() {
        return summaryKey;
    }


    public String getTitle() {
        return self.getTitle();
    }

    public int getID() {
        return self.getID();
    }

    @Override
    public void select() {
        if (!isSummary()) isSelected = !isSelected;
    }

    public PoiCategory getCategory() {
        return self;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/api/poi/PoiListItem.kt
package ch.bailu.aat_lib.api.poi

import ch.bailu.aat_lib.lib.filter_list.AbsListItem
import ch.bailu.aat_lib.lib.filter_list.KeyList
import org.mapsforge.poi.storage.PoiCategory

class PoiListItem : AbsListItem {
    val category: PoiCategory

    private val isSummary: Boolean
    private var isSelected = false

    private val keys: KeyList
    private val summaryKey: String


    constructor(category: PoiCategory) {
        isSummary = true
        this.category = category

        keys = KeyList(category.title)
        summaryKey = "_" + keys.getKey(0)
        keys.addKeys(summaryKey)
    }

    constructor(category: PoiCategory, summary: PoiListItem) {
        isSummary = false

        this.category = category

        summaryKey = summary.getSummaryKey()
        keys = KeyList(category.title)
        keys.addKeys(summaryKey)
    }

    override fun isSelected(): Boolean {
        return isSelected
    }

    override fun isSummary(): Boolean {
        return isSummary
    }

    override fun getKeys(): KeyList {
        return keys
    }

    override fun getSummaryKey(): String {
        return summaryKey
    }

    val title: String
        get() = category.title

    override fun getID(): Int {
        return category.id
    }

    override fun setSelected(selected: Boolean) {
        if (!isSummary()) isSelected = selected
    }
}
